redfish-core/lib/power_subsystem.hpp

```diff
@@ -1,20 +1,78 @@
 #pragma once
 
 #include "app.hpp"
+#include "dbus_utility.hpp"
+#include "error_messages.hpp"
 #include "logging.hpp"
 #include "query.hpp"
 #include "registries/privilege_registry.hpp"
 #include "utils/chassis_utils.hpp"
+#include "utils/dbus_utils.hpp"
 
 #include <boost/url/format.hpp>
+#include <sdbusplus/asio/property.hpp>
+#include <sdbusplus/unpack_properties.hpp>
 
+#include <cstdint>
 #include <memory>
 #include <optional>
 #include <string>
 
 namespace redfish
 {
 
+inline void getPowerSubsystemAllocation(
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    sdbusplus::asio::getAllProperties(
+        *crow::connections::systemBus, "xyz.openbmc_project.Settings",
+        "/xyz/openbmc_project/control/host0/power_cap",
+        "xyz.openbmc_project.Control.Power.Cap",
+        [asyncResp](const boost::system::error_code& ec,
+                    const dbus::utility::DBusPropertiesMap& propertiesList) {
+        if (ec)
+        {
+            if (ec.value() != EBADR)
+            {
+                BMCWEB_LOG_ERROR("DBUS response error: {}", ec.value());
+                messages::internalError(asyncResp->res);
+            }
+            return;
+        }
+
+        uint32_t powerCap{0};
+        bool powerCapEnable{false};
+        uint32_t maxPowerCapValue{0};
+        const bool success = sdbusplus::unpackPropertiesNoThrow(
+            dbus_utils::UnpackErrorPrinter(), propertiesList, "PowerCap",
+            powerCap, "PowerCapEnable", powerCapEnable, "MaxPowerCapValue",
+            maxPowerCapValue);
+
+        if (!success)
+        {
+            messages::internalError(asyncResp->res);
+            return;
+        }
+
+        // If MaxPowerCapValue valid, store Allocation properties in JSON
+        if ((maxPowerCapValue > 0) && (maxPowerCapValue < UINT32_MAX))
+        {
+            if (powerCapEnable)
+            {
+                asyncResp->res.jsonValue["Allocation"]["AllocatedWatts"] =
+                    powerCap;
+            }
+            else
+            {
+                asyncResp->res.jsonValue["Allocation"]["AllocatedWatts"] =
+                    maxPowerCapValue;
+            }
+            asyncResp->res.jsonValue["Allocation"]["RequestedWatts"] =
```

Comment:
Can you explain this mapping to me? By the PDI definition and the Redfish definition, isn't clear to me why these map together?
```
            description: The Maximum supported PowerCap setting.
```


                "RequestedWatts": {
                    "description": "The potential power, in watt units, that the subsystem requests, which might be higher than the current level being consumed because the requested power includes a budget that the subsystem wants for future use.",

Reply:
This is the same mapping that we had in 1050/1060 and is what was requested by Derek Howard (Power team lead) and Martha Broyles (OCC and Power Management team lead):

Set the Redfish properties based on properties from the D-Bus xyz.openbmc_project.Control.Power.Cap interface:
* Redfish AllocatedWatts property:
  * Set to PowerCap property if PowerCapEnable is true
  * Set to MaxPowerCapValue property if PowerCapEnable is false
* Redfish RequestedWatts property:
  * Set to MaxPowerCapValue property

Reply:
I think the idea is that MaxPowerCapValue (if it has a valid value) represents the max power that could be allocated.  If the PowerCap is enabled, then that is the amout of power that is actually available/allocated.  If the PowerCap is not enabled, then the max power is available/allocated.  RequestedWatts is set to the max power, I guess assuming we always want as much as possible.

Reply:
Ack.